DOC-446: documenting map attributes with wildcards (#416)
Certora/Documentation: docs/prover/cli/options.md

**What does it do?**
Compiles every smart contract with a different compiler executable (Solidity version or Vyper). All used contracts must be listed.

> see also See {ref}`map_attributes`

Reviewer: ```suggestion
See also See {ref}`map_attributes`
```

> See also See {ref}`map_attributes`

Reviewer: ```suggestion
See also {ref}`map_attributes`.
```

**What does it do?**
Set EVM version values when different files run with different EVM versions
Passes the value of this option as is to the Solidity compiler's option `--evm-version`.

> see also See {ref}`map_attributes`

Reviewer: ```suggestion
See also See {ref}`map_attributes`
```

> See also See {ref}`map_attributes`

Reviewer: ```suggestion
See also {ref}`map_attributes`.
```

**What does it do?**
Set optimize values when different files run with different number of runs
Passes the value of this option as is to the Solidity compiler's option `--optimize` and `--optimize-runs`.

> See also {ref}`map_attributes`

Reviewer: ```suggestion
See also {ref}`map_attributes`.
```

> see also See {ref}`map_attributes`

Reviewer: ```suggestion
See also {ref}`map_attributes`
```

**What does it do?**
This flag configures whether the Solidity compiler should enable the IR-based code generator per contract. It allows different contracts in the same project to be compiled with or without the `via-ir` option.

> see also See {ref}`map_attributes`

Reviewer: ```suggestion
See also {ref}`map_attributes`
```

**What does it do?**
This flag configures whether the Solidity compiler should enable the IR-based code generator per contract. It allows different contracts in the same project to be compiled with or without the `via-ir` option.

> See also {ref}`map_attributes`

Reviewer: ```suggestion
See also {ref}`map_attributes`.
```

# Use full path to the compiler
certoraRun Bank.sol --verify Bank:Bank.spec --vyper /usr/local/bin/vyper0.3.10
```
## `Solidity compiler map attributes`

Map attribute are used when source files are not all compiled with the same `solc` parameters.
The attributes that can be set by map attributes are:

> | Map Attribute                               | Compiler Attribute                                          |

Author: added, doe

> | Map Attribute                               | Compiler Attribute                                          |

Reviewer: the table is still here

|---------------------------------------------|-------------------------------------------------------------|
| `compiler_map`| set for each `.solc` and `.vy` file to the appropriate compiler |
| `solc_optimize_map`|  enable optimization with optionally number of expected runs |
| `solc_via_ir_map`| compile with or without IR                                  |
| `Solc_evm_version_map`| setting the target `evm` version                              |

The format of map attributes are in command line:
```sh
certoraRun ... --compiler_map A=solc7.11,C_*=solc8.24,B.sol=solc8.9,src/**/*.vy=vyper0.3.0 ...
```
The same settings in a conf file:

```sh
{
  ...
  "compiler_map": {
    "A": "solc7.11",
    "C_*": "solc8.24",
    "B.sol": "solc8.9",
    "src/**/*.vy": "vyper0.3.0"
  },
  ...
}
```

> The key of each entry is either a pattern for contract name or a path pattern. Path patterns end with: `.sol`, `.vy` or `.yul` suffix.

Reviewer: ```suggestion
The key of each entry is either a contract name pattern or a path pattern. Path patterns end with: `.sol`, `.vy` or `.yul` suffix.
```

> It is not allowed to have both the map and the non-map attribute set together (e.g., `--solc` and `--compiler_map`).

Reviewer: ```suggestion
It is not allowed to have both the map and the non-map attribute set together (e.g., `solc` and `compiler_map`).
```
They may appear in conf files.

> If a map attribute was set, all files/contracts declared in `file:` must be mapped.

Reviewer: ```suggestion
If a map attribute was set, all files/contracts declared in `files` must be mapped.
```

> For contract patterns the wildcard character `*`, replaces any character that is allowed in contract names.

Reviewer: ```suggestion
For contract patterns, the wildcard character `*` replaces any character that is allowed in contract names.
```

> For path patterns the `*` stands for any character that is allowed in paths that is not a slash(/)

Reviewer: ```suggestion
For path patterns, the `*` stands for any character that is allowed in paths that is not a slash (`/`).
```

`**` stands for any number of directories (including none).
When a map attribute is defined and the prover calls the solidity compiler, the following will take place:
- The entries of the map attributes will be checked for a match by their order of appearance.
- If the key is a path pattern, the path of the file will be matched to the path pattern; if there is a match, the value of the entry will be used.
- If the entry is a contract pattern, the prover will check if the file contains a contract that matches the contract pattern.

> - Since all contracts and files are matched as part of the validation phase, it is not expected at this stage not to find a match.

Reviewer: I think we can delete this sentence

Reviewer: It is not clear from your sentence what we do not handle

certoraRun Bank.sol --verify Bank:Bank.spec --vyper /usr/local/bin/vyper0.3.10
```

(map_attributes)=
## `Solidity compiler map attributes`

When source files are not all compiled with the same `solc` parameters you should use Solidity compiler map attributes.

| Map Attribute                               | Compiler Attribute                                          |
|---------------------------------------------|-------------------------------------------------------------|
| `compiler_map`| Set the appropriate compiler for each `.solc` and `.vy` file |
| `solc_optimize_map`| Set optimization per file with optionally the number of expected runs |
| `solc_via_ir_map`| Compile with or without IR                                  |
| `solc_evm_version_map`| Setting the target `evm` version                              |

The format of map attributes in the command line is:
```sh
certoraRun ... --compiler_map A=solc7.11,C_*=solc8.24,B.sol=solc8.9,src/**/*.vy=vyper0.3.0 ...
```
The same settings in a conf file:

```sh
{
  ...
  "compiler_map": {
    "A": "solc7.11",
    "C_*": "solc8.24",
    "B.sol": "solc8.9",
    "src/**/*.vy": "vyper0.3.0"
  },
  ...
}
```

> The key of each entry is either a contract name pattern or a path pattern. Path patterns end with: `.sol`, `.vy` or `.yul` suffix.

Reviewer: ```suggestion
The key of each entry is either a contract name pattern or a path pattern. Path patterns must end with one of the following suffixes: `.sol`, `.vy`, or `.yul`.
```

> It is not allowed to have both the map and the non-map attribute set together (e.g., `solc` and `compiler_map`).

Reviewer: ```suggestion
It is not allowed to set both the map and the non-map attributes together (e.g., {ref}`solc` and {ref}`compiler_map`).
```

> If a map attribute was set, all files/contracts declared in `files` must be mapped.

Reviewer: ```suggestion
If a map attribute was set, all files/contracts declared in {ref}`files` must be mapped.
```

certoraRun Bank.sol --verify Bank:Bank.spec --vyper /usr/local/bin/vyper0.3.10
```

(map_attributes)=

> ## `Solidity compiler map attributes`

Reviewer: ```suggestion
## Solidity compiler map attributes
```
Shouldn't be in code tags as it is not the name of a json key or a flag.

[solc_evm_version_map](#--solc_evm_version_map) and [solc_via_ir_map](#--solc_via_ir_map).

The format of map attributes in the command line is:
```sh
certoraRun ... --compiler_map A=solc7.11,C_*=solc8.24,B.sol=solc8.9,src/**/*.vy=vyper0.3.0 ...
```
The same settings in a conf file:

```sh
{
  ...
  "compiler_map": {
    "A": "solc7.11",
    "C_*": "solc8.24",
    "B.sol": "solc8.9",
    "src/**/*.vy": "vyper0.3.0"
  },
  ...
}
```
The key of each entry is either a contract name pattern or a path pattern. Path patterns must end with one of the following suffixes: `.sol`, `.vy`, or `.yul`.

It is not allowed to set both the map and the non-map attributes together (e.g., [solc](#--solc) and [compiler_map](#--compiler_map)).

If a map attribute was set, all files/contracts declared in as sources must be mapped.

For contract patterns, the wildcard character `*` replaces any character that is allowed in contract names.

For path patterns, the `*` stands for any character that is allowed in paths that is not a slash (`/`).
`**` stands for any number of directories (including none).

> When a map attribute is defined and the prover calls the solidity compiler, the following will take place:

Reviewer: ```suggestion
When a map attribute is defined and the Prover calls the Solidity compiler, the following will take place:
```

- The entries of the map attributes will be checked for a match by their order of appearance.
- If the key is a path pattern, the path of the file will be matched to the path pattern; if there is a match, the value of the entry will be used.

> - If the entry is a contract pattern, the prover will check if the file contains a contract that matches the contract pattern.

Reviewer: ```suggestion
- If the entry is a contract pattern, the Prover will check if the file contains a contract that matches the contract pattern.
```
